src/tools/scripts/set_lighthouse_urls.sh

```diff
@@ -0,0 +1,105 @@
+    case "$opt" in
+    h|\?)
+        show_help
+        exit 0
+        ;;
+    p)  production=1
+        ;;
+    esac
+done
+shift "$((OPTIND-1))" # Discard the options and sentinel --
+
+# This script must be run from src directory
+if [ -d "src" ]; then
+  cd src || exit
+fi
+
+LIGHTHOUSE_URLS=""
+
+# Set some URLs that should always be checked on pull requests
+# to ensure basic coverage
+BASE_URLS=$(cat <<-END
+http://127.0.0.1:8080/en/2019/
+http://127.0.0.1:8080/en/2019/javascript
+http://127.0.0.1:8080/en/2020/
+END
+)
+
+if [ "${production}" == "1" ]; then
+    
+    # Get the production URLs from the production sitemap
+    LIGHTHOUSE_URLS=$(curl -s https://almanac.httparchive.org/sitemap.xml | grep "<loc" | grep -v static | sed 's/ *<loc>//g' | sed 's/<\/loc>//g')
+
+    # Switch to the Production Config file
+    LIGHTHOUSE_CONFIG_FILE="${LIGHTHOUSE_PROD_CONFIG_FILE}"
+
+elif [ "${RUN_TYPE}" == "pull_request" ] && [ "${COMMIT_SHA}" != "" ]; then
+
+    # If this is part of pull request then get list of files as those changed
+    # Uses similar logic to GitHub Super Linter
+    # First checkout main to get list of differences
+    git pull --quiet
+    git checkout main
+    # Then get the changes
+    CHANGED_FILES=$(git diff --name-only "main...${COMMIT_SHA}" --diff-filter=d content templates | grep -v base.html | grep -v ejs | grep -v base_ | grep -v toc.html | grep -v sitemap)
+    # Then back to the pull request changes
+    git checkout --progress --force "${COMMIT_SHA}"
+
+    # Transform the files to http://127.0.0.1:8080 URLs
+    LIGHTHOUSE_URLS=$(echo "${CHANGED_FILES}" | sed 's/src\/content/http:\/\/127.0.0.1:8080/g' | sed 's/\.md//g' | sed 's/\/base\//\/en\//g' | sed 's/src\/templates/http:\/\/127.0.0.1:8080/g' | sed 's/\.html//g' | sed 's/_/-/g' | sed 's/\/2019\/accessibility-statement/\/accessibility-statement/g' )
+    if [ "${LIGHTHOUSE_URLS}" = "" ]; then
+        LIGHTHOUSE_URLS="${BASE_URLS}"
+    else
+        LIGHTHOUSE_URLS=$( echo -e "${LIGHTHOUSE_URLS}\n${BASE_URLS}")
+    fi
+
+else
+
+    # Else test every URL (except PDFs) in the sitemap
+    LIGHTHOUSE_URLS=$(grep loc templates/sitemap.xml | grep -v static | sed 's/ *<loc>//g' | sed 's/<\/loc>//g' | sed 's/https:\/\/almanac.httparchive.org/http:\/\/127.0.0.1:8080/g')
+
+fi
+
+echo "URLS to check:"
+echo "${LIGHTHOUSE_URLS}"
+
+# Format the URLs for the lighthouse config:
+LIGHTHOUSE_URLS=$(echo "${LIGHTHOUSE_URLS}" | sed 's/^ */          "/' | sed 's/$/",/')
+# Remove the comma on the last URL
+LIGHTHOUSE_URLS=${LIGHTHOUSE_URLS:0:${#LIGHTHOUSE_URLS}-1}
+
+# Remove the first two lines form the config
+# (we'll insert them again below when adding the URLs)
+LIGHTHOUSE_CONFIG=$(sed "1,2d" ${LIGHTHOUSE_CONFIG_FILE})
+
+# Write the new config file - including the URLs
+cat > "${LIGHTHOUSE_CONFIG_FILE}" << END_CONFIG
+{
+  "ci": {
+    "collect": {
+      "url": [
+${LIGHTHOUSE_URLS}
+      ]
+    },
+${LIGHTHOUSE_CONFIG}
+END_CONFIG
```

Comment:
This looks very brittle to me as it is dependent on how exactly certain files are formatted. How about something more programmatic like:

```suggestion
echo ${LIGHTHOUSE_URLS} | jq '. | split("\n") | map(select(length > 0))' -Rs | jq '{ci: {collect: {url: .}}} > "${LIGHTHOUSE_CONFIG_FILE}"
```

Reply:
I'm gonna go with simplicity again over robustness. This runs on every commit and we're in total control of this so happy to accept the risk.

Reply:
Testing its functionality:

```
$ BASE_URLS=$(cat <<-END
http://127.0.0.1:8080/en/2019/
http://127.0.0.1:8080/en/2019/javascript
http://127.0.0.1:8080/en/2020/
END
)

$ echo "${BASE_URLS}" | jq -Rs '. | split("\n") | map(select(length > 0))' | jq '{ci: {collect: {url: .}}}'
{
  "ci": {
    "collect": {
      "url": [
        "http://127.0.0.1:8080/en/2019/",
        "http://127.0.0.1:8080/en/2019/javascript",
        "http://127.0.0.1:8080/en/2020/"
      ]
    }
  }
}
```

Reply:
Ok but is there a way to insert it into the existing config?

The intention of this but is to have most of the config in the separate JSON file (as easier to maintain) and then insert the URLs (which will be dynamic) in the right place above those other config. It seems your suggestion just overwrites the whole config file with the URLs only and therefore loses all the other config.

Is there a way for jq to read a JSON file and then insert the url section at a specific path but maintain the rest of the JSON? Because if that could be done without too much complication then I would be open to that.

Reply:
I did not realize that there were other things in the config file and you were attempting to insert an object inside. Yes, it is possible with `jq`, I will provide the code for that in a few.

Reply:
Test:

```
$ LIGHTHOUSE_URLS=$(cat <<-END
http://127.0.0.1:8080/en/2019/
http://127.0.0.1:8080/en/2019/javascript
http://127.0.0.1:8080/en/2020/
END
)

$ echo "${LIGHTHOUSE_URLS}" | jq -Rs '. | split("\n") | map(select(length > 0))' | jq -s '.[0] * {ci: {collect: {url: .[1]}}}' lighthouse-config-prod.json -
{
  "ci": {
    "assert": {
      "assertions": {
        "categories:accessibility": [
          "error",
          {
            "minScore": 1
          }
        ],
        "categories:best-practices": [
          "error",
          {
            "minScore": 1
          }
        ],
        "categories:seo": [
          "error",
          {
            "minScore": 1
          }
        ]
      }
    },
    "collect": {
      "url": [
        "http://127.0.0.1:8080/en/2019/",
        "http://127.0.0.1:8080/en/2019/javascript",
        "http://127.0.0.1:8080/en/2020/"
      ]
    }
  }
}
```

Reply:
I have updated the suggestion above to reflect this change. Please give it a try and see if it works. If you do not want to do everything in one step, it is possible to create a temporary JSON file with the list of URIs and then merge it into the config JSON. However, I like the single command approach better as it does not create unnecessary temporary files that we will need to remove after merging.

Reply:
I get this on MacOS:

```
jq: error (at <stdin>:5): array (["http://12...) and object ({"ci":{"col...) cannot be multiplied
```

But had some other problems with jq on MacOS so will commit it and see how linux in GitHub Actions likes it.

Reply:
Same thing on linux. Trying to debug it now...

Reply:
I think it is because you might be using jq version prior to `1.4` which did not have `*` support.

Reply:
> `jq: error (at <stdin>:5): array (["http://12...) and object ({"ci":{"col...) cannot be multiplied`

How are you testing it? The way I formatted things in my examples above, I have already wrapped the array inside of the object at the nested level, so there should not be the situation of merging array with object.

Reply:
Ah works now. Didn't like the fact I was reading the file and also writing to it at the same time - which the Linter picked up to be fair.

This fixed it: https://github.com/HTTPArchive/almanac.httparchive.org/pull/1401/commits/7e5686f3db889f5a607f07a5754b016a6ce95de8

This is awesome @ibnesayeed thanks very much for your jq mastery!

Reply:
Here is another way of doing the same, but without using `*` operator:

```
$ echo "${LIGHTHOUSE_URLS}" | jq -Rs '. | split("\n") | map(select(length > 0))' | jq -s '.[0].ci.collect.url = .[1] | .[0]' lighthouse-config-prod.json -
```

Reply:
Original way worked fine with that last linting commit. Example run here: https://github.com/HTTPArchive/almanac.httparchive.org/runs/1349670806?check_suite_focus=true

Reply:
> Didn't like the fact I was reading the file and also writing to it at the same time

Oops! That's an oversight on my end. I wish `jq` had a solution like `sed -i`. :smile: